core/src/main/scala/io/projectglow/gff/GffDataSource.scala

```diff
@@ -0,0 +1,346 @@
+  private[gff] val ARRAY_DELIMITER = ","
+
+  private[gff] val csvReadOptions = Map(
+    "sep" -> COLUMN_DELIMITER,
+    "comment" -> COMMENT_IDENTIFIER,
+    "mode" -> "DROPMALFORMED",
+    "nullValue" -> NULL_IDENTIFIER
+  )
+
+  private[gff] val columnPruningConf = "spark.sql.csv.parser.columnPruning.enabled"
+
+  private[gff] val officialFieldNames = gffOfficialAttributeFields.map(_.name)
+
+  /**
+   * Infers the schema by reading the gff fileusing csv datasource and parsing the attributes fields
+   * to get official and unofficial attribute columns. Drops the original "attributes" column. Names,
+   * types and ordering of columns will be as follows:
+   * Names: All attribute fields will have names exactly as in tags.
+   * Types: Official attribute fields will have the types in
+   *       [[io.projectglow.common.FeatureSchemas.gffOfficialAttributeFields]]. Detection of
+   *       official field names among tags are case-and-underscore-insensitive. Unofficial
+   *       attribute fields will have StringType.
+   * Ordering: gffBaseSchema fields come first, followed by official attributes fields as ordered
+   *           in [[io.projectglow.common.FeatureSchemas.gffOfficialAttributeFields]], followed by
+   *           unofficial attribute fields in case-insensitive alphabetical order.
+   *
+   * @param sqlContext
+   * @param path: path to gff file or glob
+   * @return inferred schema
+   */
+  def inferSchema(sqlContext: SQLContext, path: String): StructType = {
+    val spark = sqlContext.sparkSession
+
+    val originalColumnPruning: Boolean =
+      spark.conf.get(columnPruningConf).toBoolean
+
+    // csv column pruning is turned off temporarily for the malformed rows (such as Fasta rows) to
+    // get filtered by DROPMALFORMED mode option of csv reader no matter which columns are selected
+    // in the query. See https://issues.apache.org/jira/browse/SPARK-28058
+    if (originalColumnPruning) {
+      spark.conf.set(columnPruningConf, false)
+    }
+
+    val csvDf = spark
+      .read
+      .options(csvReadOptions)
+      .schema(gffBaseSchema)
+      .csv(path)
+
+    val attributeTags = addAttributesMapColumn(csvDf)
+      .withColumn(
+        attributesMapColumnName,
+        explode(
+          map_keys(
+            col(attributesMapColumnName)
+          )
+        )
+      )
+      .agg(
+        collect_set(
+          attributesMapColumnName
+        )
+      )
+      .collect()(0)
+      .getAs[Seq[String]](0)
+      .filter(!_.isEmpty)
+
+    spark.conf.set(columnPruningConf, originalColumnPruning)
+
+    val attributeFields = attributeTags
+      .foldLeft(Seq[StructField]()) { (s, t) =>
+        val officialIdx = officialFieldNames.indexOf(normalizeString(t))
+        if (officialIdx > -1) {
+          s :+ StructField(t, gffOfficialAttributeFields(officialIdx).dataType)
+        } else {
+          s :+ StructField(t, StringType)
+        }
+      }
+      .sortBy(_.name)(FieldNameOrdering)
```

Comment:
nit: I think the custom ordering is not necessary. It might be simpler to use the natural ordering on whether the field is in the official attributes and then the name.

```scala
attributes.sortBy(field => (!officialAttributes.contains(field), field.name))
```

Reply:
Used your suggestion but with indexOf because I want the official fields to be in a set order as well.